vpr/src/pack/greedy_candidate_selector.cpp

```diff
@@ -1289,30 +1325,46 @@ PackMoleculeId GreedyCandidateSelector::get_unrelated_candidate_for_cluster_appa
                 // skip it.
                 if (!cluster_legalizer.is_molecule_compatible(mol_id, cluster_id))
                     continue;
-                // Return this molecule as the unrelated candidate.
-                return mol_id;
+
+                // If this is the best candidate we have seen so far, hold onto it.
+                // Here, we get the distance needed to move the molecule from its
+                // GP placement to the current cluster's tile.
+                t_flat_pl_loc mol_pos = get_molecule_pos(mol_id, prepacker_, appack_ctx_);
+                float mol_dist = get_manhattan_distance_to_tile(mol_pos,
+                                                                cluster_tile_loc,
+                                                                g_vpr_ctx.device().grid);
+                if (mol_dist < best_candidate_distance && mol_dist < best_distance) {
+                    best_candidate = mol_id;
+                    best_candidate_distance = mol_dist;
+                }
             }
         }
 
+        // If a candidate could be found, add it as the best found so far.
+        if (best_candidate.is_valid()) {
+            closest_compatible_molecule = best_candidate;
+            best_distance = best_candidate_distance;
+        }
+
         // Push the neighbors of the position to the queue.
         // Note: Here, we are using the manhattan distance, so we do not push
         //       the diagonals. We also want to try the direct neighbors first
         //       since they should be closer.
-        if (node_loc.x >= 1.0f)
-            search_queue.push({node_loc.x - 1, node_loc.y, node_loc.layer});
-        if (node_loc.x <= visited.dim_size(0) - 2)
-            search_queue.push({node_loc.x + 1, node_loc.y, node_loc.layer});
-        if (node_loc.y >= 1.0f)
-            search_queue.push({node_loc.x, node_loc.y - 1, node_loc.layer});
-        if (node_loc.y <= visited.dim_size(1) - 2)
-            search_queue.push({node_loc.x, node_loc.y + 1, node_loc.layer});
+        if (node_loc.x >= 1)
```

Comment:
Suggest commenting why you use visited here, and that it is essentially checking that you won't go outside the grid.